File: valdi_core/src/valdi_core/ios/swift/ValdiBridgeFunction.swift
import valdi_core

public protocol ValdiBridgeFunction {
    static var className: String { get }
    static func modulePath() -> String
    static var asyncStrictMode: Bool { get }
}

public extension ValdiBridgeFunction {
    /// When async_strict_mode is enabled, asserts that the current thread is not the main thread (to avoid ANRs).
    /// Call this at the start of resolution (e.g. init(jsRuntime:)) to avoid generating the same check in every function class.
    static func assertResolutionNotOnMainThreadIfNeeded(asyncStrictMode: Bool) {
        if asyncStrictMode {
            assert(!Thread.isMainThread,
                   "When async_strict_mode is enabled, function resolution must not be called from the main thread (to avoid ANRs). Use a background thread, the JS thread, or invokeWithJSRuntime.")
        }
    }

    static func createBridgeFunction(jsRuntime: SCValdiJSRuntime) throws -> ValdiFunction {
        return try withMarshaller { marshaller in 
            ValdiMarshallableObjectRegistry.shared.setActiveSchemeOfClassInMarshaller(className: Self.className, marshaller: marshaller)
            // Must be the error-reporting variant: pushModuleAthPath(_:in:) raises an SCValdiError
            // NSException, which Swift cannot catch.
            let index = jsRuntime.pushModule(atPath: Self.modulePath(),
                                             reportingErrorOn: OpaquePointer(marshaller.marshallerCpp))
            try marshaller.checkError()
            return try marshaller.getTypedObjectOrMapProperty(index, 0, Self.className) { try marshaller.getFunction($0) }
        }
    }
}

public func createBridgeFunctionWrapper(_ bridgeFunction: @escaping (ValdiMarshaller) throws -> Void) -> (ValdiMarshaller) -> Bool {
    return { marshaller in
        do {
            try bridgeFunction(marshaller)
            return true
        } catch let error {
            marshaller.setError(error.localizedDescription)
            return false
        }
    }
}

File: valdi/src/java/com/snap/valdi/callable/ValdiBridgeFunction.kt
package com.snap.valdi.callable

import com.snap.valdi.jsmodules.ValdiJSRuntime
import com.snap.valdi.schema.ValdiValueMarshallerRegistry
import com.snap.valdi.utils.ValdiMarshaller
import com.snap.valdi.utils.isMainThread

open class ValdiBridgeFunction {

    companion object {
        /**
         * When async_strict_mode is enabled, asserts that the current thread is not the main thread (to avoid ANRs).
         * Call this at the start of resolution (e.g. create()) to avoid generating the same check in every function class.
         */
        @JvmStatic
        fun assertResolutionNotOnMainThreadIfNeeded(asyncStrictMode: Boolean) {
            if (asyncStrictMode) {
                check(!isMainThread()) {
                    "When async_strict_mode is enabled, function resolution (create) must not be called from the main thread (to avoid ANRs). Use a background thread, the JS thread, or invokeWithJSRuntime."
                }
            }
        }

        @JvmStatic
        fun <T: ValdiBridgeFunction> createFromRuntime(runtime: ValdiJSRuntime,
                                                          cls: Class<T>,
                                                          modulePath: String): T {
            return ValdiMarshaller.use {
                val registry = ValdiValueMarshallerRegistry.shared
                registry.setActiveSchemaOfClassToMarshaller(cls, it)
                val objectIndex = runtime.pushModuleToMarshaller(modulePath, it)
                it.checkError()
                ValdiValueMarshallerRegistry.shared.unmarshallObject(cls, it, objectIndex)
            }
        }
    }

}
